aws_signing_helper/update.go

```diff
@@ -98,90 +98,92 @@ func GetCredentialsFileContents() ([]string, error) {
 func GetWriteOnlyCredentialsFile() (*os.File, error) {
 	homeDir, _ := os.UserHomeDir()
 	awsCredentialsPath := os.Getenv(AwsSharedCredentialsFileEnvVarName)
 	if awsCredentialsPath == "" {
 		awsCredentialsPath = filepath.Join(homeDir, ".aws", "credentials")
 	}
-	return os.OpenFile(awsCredentialsPath, os.O_WRONLY, 0200)
+	return os.OpenFile(awsCredentialsPath, os.O_WRONLY|os.O_TRUNC, 0200)
 }
 
 // Function to write existing credentials and newly-created credentials to a destination file
-func WriteTo(profileName string, writeLines []string, cred *TemporaryCredential) error {
+func WriteTo(profileName string, readLines []string, cred *TemporaryCredential) error {
 	destFile, err := GetWriteOnlyCredentialsFile()
 	if err != nil {
 		log.Println("unable to get write-only AWS credentials file")
 		os.Exit(1)
 	}
 	defer destFile.Close()
 
-	// Create buffered writer with a buffer of maximum size
-	destFileWriter := bufio.NewWriterSize(destFile, math.MaxInt32)
-
 	var profileExist = false
 	var profileSection = "[" + profileName + "]"
 	// A variable that checks whether or not required fields are written to the destination file
 	newCredVisit := map[string]bool{"aws_access_key_id": false, "aws_secret_access_key": false, "aws_session_token": false}
 	accessKey := "aws_access_key_id = " + cred.AccessKeyId + "\n"
 	secretKey := "aws_secret_access_key = " + cred.SecretAccessKey + "\n"
 	sessionToken := "aws_session_token = " + cred.SessionToken + "\n"
-
-	for i := 0; i < len(writeLines); i++ {
-		if !profileExist && writeLines[i] == profileSection {
-			j := i + 1
-			destFileWriter.WriteString(profileSection + "\n")
-			for ; j < len(writeLines); j++ {
+	var writeLines = make([]string, 0)
+	for readLinesIndex := 0; readLinesIndex < len(readLines); readLinesIndex++ {
+		if !profileExist && readLines[readLinesIndex] == profileSection {
+			writeLines = append(writeLines[:], profileSection+"\n")
+			readLinesIndex += 1
+			for ; readLinesIndex < len(readLines); readLinesIndex++ {
 				// If the last line of the credentials file is reached
 				// OR the next profile section is reached
-				if j == len(writeLines)-1 || strings.HasPrefix(writeLines[j], "[") {
+				if readLinesIndex == len(readLines)-1 || strings.HasPrefix(readLines[readLinesIndex], "[") {
 					if !newCredVisit["aws_access_key_id"] {
-						_, err = destFileWriter.WriteString(accessKey)
+						writeLines = append(writeLines[:], accessKey)
 					}
 					if !newCredVisit["aws_secret_access_key"] {
-						_, err = destFileWriter.WriteString(secretKey)
+						writeLines = append(writeLines[:], secretKey)
 					}
 					if !newCredVisit["aws_session_token"] {
-						_, err = destFileWriter.WriteString(sessionToken)
+						writeLines = append(writeLines[:], sessionToken)
 
 					}
-					if j == len(writeLines)-1 {
-						i = j
-					} else {
-						i = j - 1
+					if readLinesIndex != len(readLines)-1 {
+						readLinesIndex -= 1
 					}
 					profileExist = true
```

Comment:
Does this ever become `false` again?

Reply:
In this function, `profileExist` won't become `false` again. But this function is called called again when credentials need to be refreshed in the case that `update` is called as a long-running process.